datafusion/common/src/scalar.rs

```diff
@@ -1368,103 +1366,36 @@ impl ScalarValue {
-                                Expected {:?}, got {:?}",
-                            data_type, sv
-                        ),
-                    })
-                    .collect::<Result<Vec<_>>>()?,
-                )))
-            }};
-        }
-
-        macro_rules! build_array_list_string {
-            ($BUILDER:ident, $STRING_ARRAY:ident,$LIST_BUILDER:ident,$SCALAR_LIST:pat) => {{
-                let mut builder = $LIST_BUILDER::new($BUILDER::new());
-                for scalar in scalars.into_iter() {
-                    match scalar {
-                        ScalarValue::List(arr) if matches!(scalar, $SCALAR_LIST) => {
-                            // `ScalarValue::List` contains a single element `ListArray`.
-                            let list_arr = as_list_array(&arr);
-
-                            if list_arr.is_null(0) {
-                                builder.append(false);
-                                continue;
-                            }
-
-                            let string_arr = $STRING_ARRAY(list_arr.values());
-
-                            for v in string_arr.iter() {
-                                if let Some(v) = v {
-                                    builder.values().append_value(v);
-                                } else {
-                                    builder.values().append_null();
-                                }
-                            }
-                            builder.append(true);
-                        }
-                        ScalarValue::LargeList(arr) if matches!(scalar, $SCALAR_LIST) => {
-                            // `ScalarValue::List` contains a single element `ListArray`.
-                            let list_arr = as_large_list_array(&arr);
-
-                            if list_arr.is_null(0) {
-                                builder.append(false);
-                                continue;
-                            }
-
-                            let string_arr = $STRING_ARRAY(list_arr.values());
-
-                            for v in string_arr.iter() {
-                                if let Some(v) = v {
-                                    builder.values().append_value(v);
-                                } else {
-                                    builder.values().append_null();
-                                }
-                            }
-                            builder.append(true);
-                        }
-                        sv => {
-                            return _internal_err!(
-                                "Inconsistent types in ScalarValue::iter_to_array. \
-                                    Expected List, got {:?}",
-                                sv
-                            )
-                        }
-                    }
+        fn build_list_array(
+            scalars: impl IntoIterator<Item = ScalarValue>,
+        ) -> Result<ArrayRef> {
+            let arrays = scalars
```

Comment:
Is it necessary to make a 1 element array for each ScalarValue?

I wonder if you could make the values array directly like

```rust
let values = ScalarValue::iter_to_array(scalars);
// and then make a single offset of 0
```

🤔

Reply:
After iter_to_array, the values is not guaranteed to be single element, so I don't think it works

Reply:
Since the result of iter_to_array is not guaranteed to be single array like what ScalarValue is, so recursive calls in this function does not make sense to me.

Reply:
I see what you are saying -- sorry I was confused.